mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+      map = loadOp.getAffineMap();
+    else if (AffineStoreOp storeOp = dyn_cast<AffineStoreOp>(srcAccess.opInst))
+      map = storeOp.getAffineMap();
+
+    SmallVector<Value, 8> operands(srcAccess.indices.begin(),
+                                   srcAccess.indices.end());
+
+    fullyComposeAffineMapAndOperands(&map, &operands);
+    map = simplifyAffineMap(map);
+    canonicalizeMapAndOperands(&map, &operands);
+
+    // Parse each map result to construct access matrices.
+    ArrayRef<AffineExpr> mapResults = map.getResults();
+
+    uint64_t noDims = map.getNumDims();
+    uint64_t noSymbols = map.getNumSymbols();
+
+    // Number of rows in an accessMatrix = Number of dimensions
+    // in the memref object.
+    // Number of Columns = (noDims + noSymbols).
+
+    SmallVector<SmallVector<int64_t, 4>, 4> accessMatrix;
+    SmallVector<int64_t, 4> constVector(mapResults.size());
+
+    for (unsigned l = 0; l < mapResults.size(); l++) {
+
+      // Parse the l-th map result (access expr for l-th dimension of this
+      // memref) to get the l-th row of this access matrix.
+      AffineExpr mapResult = mapResults[l];
+
+      // A Row of Matrix A.
+      SmallVector<int64_t, 4> Row;
+      Row.resize(
+          std::max(AffineForOpLoopNestSize, unsigned(noDims + noSymbols)));
+
+      // Check if `mapResult` is not a constant expr.
+      // If it is a constant expr like A[5], no need to walk it.
+      // Instead, push the value in constVector and
+      // leave the Row matrix to be a vector of zeroes.
+
+      if (mapResult.getKind() == AffineExprKind::Constant) {
+        auto constant = mapResult.cast<AffineConstantExpr>();
+        constVector[l] = (constant.getValue());
+      } else {
+
+        // Start parsing the mapResult.
+        mapResult.walk([&](AffineExpr expr) {
+          // Each `expr` can in turn be a combination of other affine
+          // expressions.
+          switch (expr.getKind()) {
+          case AffineExprKind::Add: {
+
+            // Does this row need a modification?
+            // Please note that the Row should not
+            // be modified in case of a constant affine expr.
+            bool modifyRow = true;
+
+            AffineBinaryOpExpr addExpr = expr.cast<AffineBinaryOpExpr>();
+            AffineExpr lhs = addExpr.getLHS();
+            AffineExpr rhs = addExpr.getRHS();
+            unsigned lhsPosition = 0;
+            unsigned rhsPosition = 0;
+
+            // Parse LHS part of affine access expr.
+            AffineExprKind lhskind = lhs.getKind();
+            if (lhskind == AffineExprKind::DimId) {
+              auto dim = lhs.cast<AffineDimExpr>();
+              lhsPosition = indVarIdMap[operands[dim.getPosition()]];
+            } else if (lhskind == AffineExprKind::SymbolId) {
+              auto symbol = lhs.cast<AffineSymbolExpr>();
+              lhsPosition = noDims + symbol.getPosition();
+            } else if (lhskind == AffineExprKind::Constant) {
```

Comment:
Don't compare with ==. Instead use isa<..> unless you are using a switch.

Reply:
Replaced with switch.